norfair/metrics.py

```diff
@@ -0,0 +1,327 @@
+
+        file_name = os.path.split(input_path)[1]
+
+        if information_file is None:
+            seqinfo_path = os.path.join(input_path, "seqinfo.ini")
+            information_file = InformationFile(file_path = seqinfo_path)
+
+        self.length = information_file.search(variable_name = "seqLength")
+
+        predictions_folder = os.path.join(save_path, "predictions")
+        if not os.path.exists(predictions_folder):
+            os.makedirs(predictions_folder)
+
+        out_file_name = os.path.join(predictions_folder, file_name + ".txt")
+        self.text_file = open(out_file_name, "w+")
+
+        self.frame_number = 1
+
+    def update_text_file(self, predictions):
+        write_predictions(
+            frame_number=self.frame_number, objects=predictions, output_text_file=self.text_file
+        )
+        self.frame_number += 1
+
+        if self.frame_number > self.length:
+            self.text_file.close()
+
+def get_detections(input_path, information_file=None):
+    class DetectionFileParser:
+        """Get Norfair detections from MOTChallenge text files containing detections"""
+        def __init__(self, input_path, information_file=None):
+            #Get detecions matrix data with rows corresponding to:
+            # frame, id, bb_left, bb_top, bb_right, bb_down, conf, x, y, z
+            detections_path = os.path.join(input_path, "det/det.txt")
+            self.matrix_detections = process_text_file(text_file_path=detections_path)
+
+            if information_file is None:
+                seqinfo_path = os.path.join(input_path, "seqinfo.ini")
+                information_file = InformationFile(file_path = seqinfo_path)
+            length = information_file.search(variable_name = "seqLength")
+
+            self.ordered_by_frame = []
+
+            for frame_number in np.arange(1, length + 1):
+                dets_on_this_frame = self.get_dets_from_frame(frame_number)
+                self.ordered_by_frame.append(dets_on_this_frame)
+
+        def get_dets_from_frame(self, frame_number):
+            """ this function returns a list of norfair Detections class, corresponding to frame=frame_number """
+
+            indexes = np.argwhere(self.matrix_detections[:, 0] == frame_number)
+            detections = []
+            if len(indexes) > 0:
+                actual_det = self.matrix_detections[indexes]
+                actual_det.shape = [actual_det.shape[0], actual_det.shape[2]]
+                for det in actual_det:
+                    points = np.array([[det[2], det[3]], [det[4], det[5]]])
+                    conf = det[6]
+                    new_detection = Detection(
+                        points, np.array([conf, conf])
+                    ) 
+                    detections.append(new_detection)
+            self.actual_detections = detections
+            return detections
+
+    return DetectionFileParser(input_path=input_path, information_file=information_file).ordered_by_frame
```

Comment:
It doesn't make sense to define a class inside a function and then instantiate it and return it. We should just define the class by itself and instantiate it when we use it in `motmetrics4norfair.py`.

What I was refering to in my last comment, was that we should not instantiate the class and immediately get its `ordered_by_frame` list. We should actually make this class iterable so we can pass it to the for loop directly without having to access `ordered_by_frame`.

I haven't run this code, but in the state it is currently in I don't think it should be able to run. The for loop calling an instance of this class in `motmetrics4norfair.py ` should actually crash as it is not iterable.

Please contact me when you reach this so we can discuss it further.